### ISSUE TITLE
Project: Finding email in database

### ISSUE THREAD

**Issue by PaulFarault** [adaltas/ece-2020-fall-webtech#14]
What is the most efficient way to find an email into our level db database as keys are generated id ? I was thinking about different possibilities :
- Reading the whole list of users and then creating a loop to find the email (not efficient ?)
- Creating a new section on the database with email as key and user_id as value
- Using email as id/key (easier but make it harder to change the email ?)

Is there another(s) way(s) to accomplish this task ?
What is the advantage of using a key->value database instead of a document one in our case (speed/efficiency ?) ?

**Comment by wdavidw** [comment]
> Reading the whole list of users and then creating a loop to find the email

Not scalable, you are right

> Creating a new section on the database with email as key and user_id as value

This is the best approach, it is called a secondary index. RDBMS usually do it for you transparently. In many NoSQL stores (Redis, LevelDB, HBase, ...), you will have to manage them yourself and keep the database consistent. For example, when the email is stored in a secondary index, removing a user implies reading his email and removing the email from secondary index before removing the user. It is more work and also more control. As Voltaire once said, with great power comes great responsibilities.

> Using email as id/key (easier but make it harder to change the email ?)

You spotted the problem, if we use the email as an identifier, we can no longer change the email. Choose only this approach if you failed at implementing the secondary index (which is not hard, just more work).

**Comment by PaulFarault** [comment]
Is there any tool that could be used to visualize the whole level database for testing purposes ?

**Comment by wdavidw** [comment]
not that I know but you could create a script which traverses the all database and prints its content

**Comment by PaulFarault** [comment]
> In many NoSQL stores (Redis, LevelDB, HBase, ...), you will have to manage them yourself and keep the database consistent. For example, when the email is stored in a secondary index, removing a user implies reading his email and removing the email from secondary index before removing the user. It is more work and also more control.

I'm going to give a try to this approach. However, it raises some new questions as I don't really know how to work with a secondary index.

### Using a secondary index
- Should the email be stored on both indexes or only on the secondary one ?
- When should I use the secondary index :
    - Only the first time to get the id (and of course to modify the email) and then only work with the id and primary index.
    - As a "main entry", to handle most of my API calls. _I don't think that this proposition is very clever as it seems to be the same thing as using email as id but who knows._

### Defining API endpoints
In general, is it better to have distinct endpoints for each purpose or should we build "smarter' endpoints ?
- Example on secondary index : 
    - 1 endpoint to get id from email + 1 endpoint to get user info from id
    - OR 1 single endpoint to get all user info from email

- Example on creating a new user : 
    - 1 endpoint to try to get the user + 1 endpoint to create a new user
    - OR 1 single get endpoint which create the user if it doesn't exist

Thank you for your help !

**Comment by wdavidw** [comment]
> Should the email be stored on both indexes or only on the secondary one ?

You need to store the email with the user data, otherwise you won't be able to get it when you get a user from its user id. The secondary index is just a place where you associate email with id, kind of like a pointer un programming. It involves 2 lookups: 1, for a given email, get the user id; 2, for a given user id, get the user object. It is better to access directly the user object with only one lookup when you have his user id but this is not always possible and it is much better to do 2 lookups than to traverse the all database which is never an option.

**Comment by wdavidw** [comment]
You use the secondary index when you don't have the choice, which is when you don't have the user id and only the user email, which is your case with the access token. This is very quick because the type of database we are using are very fast. If RDMS are slower, it is because they do the same things behind the scene. You create create HTTP endpoints for this, it would be slow and useless, you must handle this in your app or db layer. Both are possible, I'll probably do it in the app layer.

**Comment by PaulFarault** [comment]
> You dont create/modify any HTTP endpoints for this, it would be slow and useless

Shouldn't I still need to create 1 endpoint to get id from email ?

**Comment by wdavidw** [comment]
If by endpoint you mean HTTP endpoint, I don't see the use for it.

**Comment by PaulFarault** [comment]
Maybe something that looks more like this ?
``` js
app.get('users/:id', (req, res) => {
  if (req.params.id == 0) {
    // get mail from req.body
    // look for id in secondary index
  }
  // look for user object in primary index from id
  // res.send user object
}
```

**Comment by wdavidw** [comment]
Why to make things so complicated, `app.get('users/:id')`, you have the id, just query LevelDB with `db.get(`channels:${id}`)` and done. Why would you need the email, there is no need for a second index here. I don't get it. The final route would look like:

```js
app.get('/users/:id', authenticate, async (req, res) => {
  const user = await db.users.get(req.params.id)
  // make sure a user only access his personal data
  // the authenticate middleware has created
  // `req.user` with the JWT payload, let's use it
  if(user.email !== req.user.email){
      res.status(418).send('Oh no')
      return
  }
  res.json(user)
})
```

**Comment by PaulFarault** [comment]
Thanks for this answer, I think that now I understand the purpose of both indexes. My main trouble is about _Where can I ask for my id by giving my email ?_

It should be a place where I can send an HTTP request with my email in order to get my id and then accomplish the rest of my requests. Which endpoint should I use ?

If I don't make mistake, your piece of code only allow me to get an user object with an id ?

**Comment by wdavidw** [comment]
The code I provided reflect the use case you provided and how I would implement it. Let's say you want to display the channels of a user, it would have been different. the endpoints doesn't change, `app.get('/channels')`. What change is that now, you have the verified identify of who issue the request, thanks to the OAuth and `authenticate` middleware, and the endpoint must only return the channel of the provided identity. For example:

```js
app.get('/channels', authenticate, async (req, res) => {
  // The authenticate middleware has enriched `req` with the `user` object
  const {id: userId} = await db.users.getByEmail(req.user.email)
  const channels = await db.channels.listForUser(userId)
  res.json(channels)
})
```

How does `getByEmail`, maybe like:

```
{
  "users": {
    getByEmail: async (email) => {
      const id = await db.get(`users_by_email:${emails}`)
      return module.exports.users.get(id)
    },
  }
}
```

**Comment by PaulFarault** [comment]
Thank you for this final answer ! It was very clear and answer all of my questions.
(except for the `return module.exports.users.get(id)` part, I'm not really sur about what it's doing)

One more last thing : it seems that I can't use a body for requests which are using `authenticate` in the header. Is this a problem only for me or in general ? (maybe it's normal ?) 

``` js
// 1st method
const { data: res } = await axios({
  method: 'get',
  url: 'http://localhost:3001/users/0',
  data: { data: "my data" },
  headers: {
    Authorization: 'Bearer ' + oauth.access_token,
    UserId: user.id
  }
})

// 2nd method
const {data: channels} = await axios.get('http://localhost:3001/channels',
  { data: "my data" },
  {headers: {
    'Authorization': `Bearer ${oauth.access_token}`,
    'UserId': user.id
  }}
)
```
In both cases on the back-end I'm able to grab `userId` using `req.headers.userid` but `req.body` shows an empty object `{}`. Is that normal ?
Maybe I'm doing a misuse of the `authenticate` middleware ?

**Comment by wdavidw** [comment]
You should not send UserId in the HTTP request, you already know the identity of your user, it is the email inside the access token.

**Comment by PaulFarault** [comment]
Oh ok !

> You use the secondary index when you don't have the choice, which is when you don't have the user id and only the user email, which is your case with the access token.

I thought that you where saying that in the case of a single application. I interpreted it as "at the beginning we don't have userId, only email, so try to get userId to use it for your requests". That's why I proposed to create some functions to get bask the user id !

So, if I finally understand, in our **whole** application we should only use the `authenticate` middleware and never send the userId.
